### PR TITLE
Layer order

### PR DESCRIPTION
**Pull request by abbytiz**
This PR uses a python script to unpack the layers in order resulting in the creation of the final file system of the image.  In order to make this switch, all differs had to be altered to take in a final file-system rather than the layer structure previously used.  There is an additional integration test added which tests the layer order handling.  Additionally, in unpacking to the final file system it became apparent that the pip differ needed to switch to the multi version package handler because if there are multiple versions of python installed, multiple versions of a package could also be installed.

Currently there are merge conflicts that will be resolved in the next commit, but the overall intent of the PR is able to begin the review process.

@cftorres @aaron-prindle @nkubala 
#263

### CODE REVIEW

File: iDiff/differs/pipDiff.go [hunk 1]
```diff
@@ -14,71 +14,77 @@ type PipDiffer struct {
...
+		// layer doesn't have a Python version installed
+		return packages, nil
+	}
+	for _, pyVersion := range pythonVersions {
+		packagesPath := filepath.Join(path, "usr/local/lib", pyVersion, "site-packages")
 		contents, err := ioutil.ReadDir(packagesPath)
 		if err != nil {
 			// layer's Python folder doesn't have a site-packages folder
-			continue
+			return packages, nil
```
Comment (nkubala) [inline iDiff/differs/pipDiff.go]: isn't it possible that a layer has Python installed, but no packages in the site-packages folder? the two potential cases I can think of are either no packages installed period, or packages installed in a different location (and then added to the `PYTHONPATH` manually, unlikely but still possible)

File: iDiff/undocker.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    p.set_defaults(level=logging.WARN)
+    return p.parse_args()
+
+
+def main():
+    args = parse_args()
+    logging.basicConfig(level=args.loglevel)
+    # open tar file
+    img = tarfile.open(args.tar)
+    print args.tar
```
Comment (nkubala) [inline iDiff/undocker.py]: remove?

File: iDiff/undocker.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    # extract manifest file from tar to read layers
+    repos = img.extractfile('manifest.json')
```
Comment (nkubala) [inline iDiff/undocker.py]: probably should check to be sure the manifest file exists? are there cases where it could be absent?

File: iDiff/undocker.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    repos = json.load(repos)
+    layers = repos[0]["Layers"]
+
+    # if args.layers:
+    #     sys.exit(0)
```
Comment (nkubala) [inline iDiff/undocker.py]: remove

File: iDiff/undocker.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+    if not os.path.isdir(args.output):
+        os.mkdir(args.output)
+
+    # removelist = []
```
Comment (nkubala) [inline iDiff/undocker.py]: remove

File: iDiff/undocker.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+                            shutil.rmtree(args.output + "/" + newpath)
+                            # removelist.append(newpath)
+                            # removelist.append(path)
+                        except OSError as err:
+                            LOG.info('error %s', err)
+                            if err.errno != errno.ENOENT:
+                                raise
+    img.close()
+    # for path in removelist:
+    #     shutil.rmtree(args.output + "/" + path)
```
Comment (nkubala) [inline iDiff/undocker.py]: remove

File: iDiff/undocker.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+                        else:
+                            newpath = path.replace('/.wh.', '/')
+                        try:
+                            LOG.info('removing path %s', newpath)
+                            # os.unlink(path)
+                            # os.unlink(newpath)
+                            os.remove(args.output + "/" + path)
+                            shutil.rmtree(args.output + "/" + newpath)
+                            # removelist.append(newpath)
+                            # removelist.append(path)
```
Comment (nkubala) [inline iDiff/undocker.py]: remove?

File: iDiff/undocker.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+                fileobj=img.extractfile(id),
+                errorlevel=(0 if args.ignore_errors else 1)) as layer:
+            layer.extractall(path=args.output)
+            if not args.no_whiteouts:
+                LOG.info('processing whiteouts')
+                for member in layer.getmembers():
+                    path = member.path
+                    if path.startswith('.wh.') or '/.wh.' in path:
+                        if path.startswith('.wh.'):
+                            newpath = path[4:]
```
Comment (nkubala) [inline iDiff/undocker.py]: can you use `path.replace()` here as well? and just replace with whitespace

File: iDiff/differs/pipDiff.go (hunk 1 above)
Comment (abbytiz) [inline iDiff/differs/pipDiff.go]: @cftorres do you have commentary on this?  My understanding was that the pip differ only supported python packages in the site-packages folder.

File: iDiff/differs/aptDiff.go [hunk 9]
```diff
@@ -20,29 +21,30 @@ func (d AptDiffer) Diff(image1, image2 utils.Image) (utils.DiffResult, error) {
 
 func (d AptDiffer) getPackages(path string) (map[string]utils.PackageInfo, error) {
 	packages := make(map[string]utils.PackageInfo)
-	layerStems, err := utils.BuildLayerTargets(path, "layer/var/lib/dpkg/status")
+	_, err := os.Stat(path)
```
Comment (abbytiz) [inline iDiff/differs/aptDiff.go]: done

File: iDiff/differs/fileDiff.go [hunk 10]
```diff
@@ -72,44 +57,28 @@ func diffImageFiles(image1, image2 utils.Image) (utils.DirDiff, error) {
...
-			return contents, fmt.Errorf("Could not get Directory struct for layer %s in image %s: %s", layer, pathToImage, err)
-		}
-		contents[layer] = layerDir
+	layerDir, err := utils.GetDirectory(pathToJSON)
+	defer os.Remove(pathToJSON)
+	if err != nil {
+		return contents, fmt.Errorf("Could not get Directory struct for layer %s in image %s: %s", pathToImage, pathToImage, err)
 	}
+	contents[pathToImage] = layerDir
+	// }
```
Comment (abbytiz) [inline iDiff/differs/fileDiff.go]: done

File: iDiff/undocker.py (hunk 2 above)
Comment (abbytiz) [inline iDiff/undocker.py]: done

File: iDiff/undocker.py (hunk 8 above)
Comment (abbytiz) [inline iDiff/undocker.py]: done

File: iDiff/undocker.py (hunk 5 above)
Comment (abbytiz) [inline iDiff/undocker.py]: done

File: iDiff/undocker.py (hunk 4 above)
Comment (abbytiz) [inline iDiff/undocker.py]: done

File: iDiff/undocker.py (hunk 3 above)
Comment (abbytiz) [inline iDiff/undocker.py]: This python script will be replaced in my next PR with go code when the docker dependency is removed.  In this current state of code, unless a .tar file without a manifest is passed, the manifest file is guaranteed to exist.  I did add a check though just in case.

File: iDiff/undocker.py (hunk 7 above)
Comment (abbytiz) [inline iDiff/undocker.py]: done

File: iDiff/undocker.py (hunk 6 above)
Comment (abbytiz) [inline iDiff/undocker.py]: done

File: iDiff/differs/pipDiff.go (hunk 1 above)
Comment (nkubala) [inline iDiff/differs/pipDiff.go]: by default, pip just puts all packages in site-packages. but you can configure it to install them anywhere, for example by passing `--target=/some/random/directory` to `pip install` and then adding that directory to your `PYTHONPATH`. it might be the case that we want to check all directories on the `PYTHONPATH` when we do the python diff.

check out [this StackOverflow post](https://stackoverflow.com/questions/2915471/install-a-python-package-into-a-different-directory-using-pip) for more info on this

Comment (abbytiz) [inline iDiff/differs/pipDiff.go]: Is that something you think should go in this PR or would it make sense to have that separate?

Comment (nkubala) [inline iDiff/differs/pipDiff.go]: I think it's fine to do in a separate PR. https://github.com/GoogleCloudPlatform/runtimes-common/issues/299

File: iDiff/differs/fileDiff.go [hunk 11]
```diff
@@ -33,20 +32,6 @@ func diffImageFiles(image1, image2 utils.Image) (utils.DirDiff, error) {
...
-			}
-		}
-		if sameLayer {
-			delete(img1Contents, layer1)
-		}
-	}
-
 	img1Dir := utils.Directory{
 		Root:    img1,
 		Content: getContentList(img1Contents),
```
Comment (cftorres) [inline iDiff/differs/fileDiff.go]: just turn the image FS into a Directory struct immediately from its path without all this extra content getting stuff

File: iDiff/utils/image_prep_utils.go [hunk 12]
```diff
@@ -115,11 +117,18 @@ func getHistory(imgPath string) ([]string, error) {
 
 func getImageFromTar(tarPath string) (string, error) {
 	glog.Info("Extracting image tar to obtain image file system")
-	err := ExtractTar(tarPath)
-	if err != nil {
-		return "", err
-	}
 	path := strings.TrimSuffix(tarPath, filepath.Ext(tarPath))
+	args := []string{"iDiff/undocker.py", "--tar", tarPath, "-o", path, "-v"}
+	undockerCmd := exec.Command("python", args...)
```
Comment (aaron-prindle) [inline iDiff/utils/image_prep_utils.go]: Ideally "undockering" the files will be done in native Go Code.  Understandably porting the code over costs time, I personally still think this is worthwhile as it appears there is not too much functional code that would need to be ported:
https://github.com/larsks/undocker/blob/master/undocker.py.   If absolutely necessary, the python file can be stored inside of the packaged go binary using go-bindata: https://github.com/jteeuwen/go-bindata.  "undocker" does not seem to require any additional packages so no additional setup should be required.  Python is standard in most Linux distros, this won't work out of the box though on Windows, I'm not sure about OSX. Also this python 2 specific code though, I don't think it will work with Python 3.  With this in mind, porting the code over would be ideal but we can definitely discuss this.  Also I don't think that the undocker.py file is currently in the PR.

Comment (abbytiz) [inline iDiff/utils/image_prep_utils.go]: I already have the code translated to Go in another branch that was intended to be merged after this PR went through.  I'm now working through trying to work it in to get the tests passing.  Also I see the undocker.py file in the "Files changed" Tab so I think it is added in the PR... (also see undocker.py here: https://github.com/abbytiz/runtimes-common/blob/e0e40a3e3829ce29df26e7ee6a6067cc0980d2c7/iDiff/undocker.py)

**Comment by aaron-prindle** [comment]
#304 implements the feature now, closing
